diagnostics/test/diagnostics_tests.cpp

```diff
@@ -0,0 +1,117 @@
+
+static const std::string async_handler_name2("async_test_handler2");
+
+static const std::string keylist = async_handler_name + "\n" + async_handler_name2 + "\n" + sync_handler_name + "\n";
+
+StatusHandler sync_handler(sync_handler_name, sync_handler_description, []() { return sync_handler_status; });
+
+StatusHandler async_handler(async_handler_name, async_handler_description, []() {
+  // This mimics some arbitrary asynch behavior, such as sending an internal message containing a
+  // promise to the replica thread, and then waiting on the future.
+  std::promise<std::string> promise;
+  auto future = promise.get_future();
+  std::thread t([&promise]() { promise.set_value(async_handler_status); });
+  t.join();
+  return future.get();
+});
+
+StatusHandler async_handler2(async_handler_name2, async_handler_description, []() {
+  // This mimics some arbitrary asynch behavior, such as sending an internal message containing a
+  // promise to the replica thread, and then waiting on the future.
+  std::promise<std::string> promise;
+  auto future = promise.get_future();
+  std::async([promise = std::move(promise)]() mutable { promise.set_value(async_handler_status); });
+  return future.get();
+});
+
+TEST(diagnostics_tests, status_registration) {
+  Registrar registrar;
+  registrar.registerStatusHandler(sync_handler);
+  registrar.registerStatusHandler(async_handler);
+  registrar.registerStatusHandler(async_handler2);
+  ASSERT_EQ(sync_handler_status, registrar.getStatus(sync_handler_name));
+  ASSERT_EQ(sync_handler_description, registrar.describeStatus(sync_handler_name));
+  ASSERT_EQ(async_handler_status, registrar.getStatus(async_handler_name));
+  ASSERT_EQ(async_handler_description, registrar.describeStatus(async_handler_name));
+  ASSERT_EQ(async_handler_status, registrar.getStatus(async_handler_name2));
+  ASSERT_EQ(async_handler_description, registrar.describeStatus(async_handler_name2));
+
+  ASSERT_EQ("*--STATUS_NOT_FOUND--*", registrar.getStatus("no such handler"));
+  ASSERT_EQ("*--DESCRIPTION_NOT_FOUND--*", registrar.describeStatus("no such handler"));
+
+  ASSERT_EQ(keylist, registrar.listStatusKeys());
+  std::cout << registrar.describeStatus() << std::endl;
+  std::cout << registrar.listStatusKeys() << std::endl;
+}
+
+// This tests is really just for visual confirmation, since strings are always returned.
+TEST(diagnostics_tests, protocol) {
+  Registrar registrar;
+  registrar.registerStatusHandler(sync_handler);
+  registrar.registerStatusHandler(async_handler);
+  registrar.registerStatusHandler(async_handler2);
+
+  // No parameters trigger usage to be displayed.
+  ASSERT_EQ(0, std::memcmp("Usage:", run({}, registrar).c_str(), 6));
+
+  // Bad subject triggers usage
+  ASSERT_EQ(0, std::memcmp("Usage:", run({"no_such_subject"}, registrar).c_str(), 6));
+
+  // Good subject, bad command triggers usage
+  ASSERT_EQ(0, std::memcmp("Usage:", run({"status", "bad_command"}, registrar).c_str(), 6));
```

Comment:
[Minor]: I think we can use string literals, i.e. `"Usage"s` to avoid specifying the length.

Reply:
I'm not sure I follow. I just want to compare that "Usage:" is a prefix of the output from run.

Reply:
Again, I should stop doing reviews early in the morning... :) My thought was to avoid `std::memcmp()`. I guess could be done by `run(...).find("Usage:") == 0`, but you can safely ignore that comment, not a big deal.